File: Fbs.WebApi/Repository/UserRepository.cs
using System.Linq.Expressions;
using Fbs.WebApi.Entities;
using Fbs.WebApi.Options;
using Google.Apis.Sheets.v4;
using Google.Apis.Sheets.v4.Data;
using Microsoft.Extensions.Caching.Hybrid;
using Microsoft.Extensions.Options;

namespace Fbs.WebApi.Repository;

public class UserRepository(
    InstrumentationSource instrumentation,
    IOptions<GoogleOptions> options,
    HybridCache cache,
    SheetsService sheetsService
) : IRepository<User>
{
    private readonly string[] _header =
    [
        "Unit", "Name", "Phone", "TelegramChatId", "NotificationGroup"
    ];

    public async Task<List<User>> GetListAsync(CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await cache.GetOrCreateAsync(
            "Users",
            (sheetsService, options),
            async (state, ct) =>
                await sheetsService.Spreadsheets.Values.Get(options.Value.SpreadsheetId, "Users")
                    .ExecuteAsync(ct),
            cancellationToken: cancellationToken);

        if (!items.Values.First().SequenceEqual(_header))
        {
            throw new Exception("Unexpected headers for Users sheet");
        }

        return items.Values
            .Skip(1)
            .Select((row, idx) => new User
            {
                Row = idx + 2,
                Unit = row.ElementAtOrDefault(0) as string,
                Name = row.ElementAtOrDefault(1) as string,
                Phone = row.ElementAtOrDefault(2) as string,
                TelegramChatId = row.ElementAtOrDefault(3) as string,
                NotificationGroup = row.ElementAtOrDefault(4) as string,
            })
            .ToList();
    }

    public async Task<User?> FindAsync(Expression<Func<User, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await GetListAsync(cancellationToken);
        return items.SingleOrDefault(predicate.Compile());
    }

    public async Task<User> GetAsync(Expression<Func<User, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await GetListAsync(cancellationToken);
        return items.Single(predicate.Compile());
    }

    public async Task<User> InsertAsync(User entity, CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException();
    }

    public async Task<User> UpdateAsync(User entity, CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var request = sheetsService.Spreadsheets.Values.Update(
            new ValueRange
            {
                Values = [[entity.Unit, entity.Name, entity.Phone, entity.TelegramChatId, entity.NotificationGroup]]
            },
            options.Value.SpreadsheetId,
            $"Users!A{entity.Row}:E{entity.Row}"
        );

        request.ValueInputOption = SpreadsheetsResource.ValuesResource.UpdateRequest.ValueInputOptionEnum.RAW;
        await request.ExecuteAsync(cancellationToken);

        await cache.RemoveAsync("Users", cancellationToken);

        return entity;
    }

    public async Task DeleteAsync(Expression<Func<User, bool>> predicate, CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();
        var entity = await GetAsync(predicate, cancellationToken);

        await sheetsService.Spreadsheets.BatchUpdate(new BatchUpdateSpreadsheetRequest
        {
            Requests =
            [
                new Request
                {
                    DeleteDimension = new DeleteDimensionRequest
                    {
                        Range = new DimensionRange
                        {
                            SheetId = await GetSheetId(cancellationToken),
                            Dimension = "ROWS",
                            StartIndex = entity.Row - 1,
                            EndIndex = entity.Row
                        }
                    }
                }
            ]
        }, options.Value.SpreadsheetId).ExecuteAsync(cancellationToken);
        
        await cache.RemoveAsync("Users", cancellationToken);
    }

    private async Task<int?> GetSheetId(CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var sheet = await sheetsService.Spreadsheets.Get(options.Value.SpreadsheetId).ExecuteAsync(cancellationToken);
        return sheet.Sheets.Single(s => s.Properties.Title == "Users").Properties.SheetId;
    }
}

File: Fbs.WebApi/src/main/kotlin/sg/from/fbs/repository/UserRepository.kt
package sg.from.fbs.repository

import com.google.api.services.sheets.v4.Sheets
import com.google.api.services.sheets.v4.model.*
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Repository
import sg.from.fbs.config.GoogleProperties
import sg.from.fbs.entities.User
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Lazy

@Repository
class UserRepository(
    private val sheetsService: Sheets,
    private val options: GoogleProperties
) {
    @Autowired
    @Lazy
    lateinit var self: UserRepository

    private val header = listOf("Unit", "Name", "Phone", "TelegramChatId", "NotificationGroup", "IsAdmin")

    @Cacheable("Users")
    fun getList(): List<User> {
        val response = sheetsService.spreadsheets().values()
            .get(options.spreadsheetId, "Users")
            .execute()

        val values = response.getValues()
        if (values == null || values.isEmpty()) return emptyList()

        val actualHeader = values.first().map { it.toString() }
        if (actualHeader != header) {
            throw Exception("Unexpected headers for Users sheet")
        }

        return values.drop(1).mapIndexed { idx, row ->
            User(
                row = idx + 2,
                unit = row.getOrNull(0) as? String,
                name = row.getOrNull(1) as? String,
                phone = row.getOrNull(2) as? String,
                telegramChatId = row.getOrNull(3) as? String,
                notificationGroup = row.getOrNull(4) as? String,
                isAdmin = (row.getOrNull(5) as? String) == "TRUE"
            )
        }
    }

    fun find(predicate: (User) -> Boolean): User? {
        return self.getList().find(predicate)
    }

    fun get(predicate: (User) -> Boolean): User {
        return self.getList().first(predicate)
    }

    fun insert(entity: User): User {
        throw NotImplementedError()
    }

    @CacheEvict(value = ["Users"], allEntries = true)
    fun update(entity: User): User {
        val body = ValueRange().setValues(
            listOf(
                listOf(
                    entity.unit,
                    entity.name,
                    entity.phone,
                    entity.telegramChatId,
                    entity.notificationGroup,
                    entity.isAdmin.toString().uppercase()
                )
            )
        )

        sheetsService.spreadsheets().values()
            .update(options.spreadsheetId, "Users!A${entity.row}:F${entity.row}", body)
            .setValueInputOption("RAW")
            .execute()

        return entity
    }

    @CacheEvict(value = ["Users"], allEntries = true)
    fun delete(predicate: (User) -> Boolean) {
        val entity = get(predicate)
        val sheetId = getSheetId()

        val request = Request().setDeleteDimension(
            DeleteDimensionRequest().setRange(
                DimensionRange()
                    .setSheetId(sheetId)
                    .setDimension("ROWS")
                    .setStartIndex(entity.row - 1)
                    .setEndIndex(entity.row)
            )
        )

        sheetsService.spreadsheets().batchUpdate(
            options.spreadsheetId,
            BatchUpdateSpreadsheetRequest().setRequests(listOf(request))
        ).execute()
    }

    private fun getSheetId(): Int {
        val spreadsheet = sheetsService.spreadsheets().get(options.spreadsheetId).execute()
        val sheet = spreadsheet.sheets.single { it.properties.title == "Users" }
        return sheet.properties.sheetId
    }
}